src/kernel/txgraph.cpp

```diff
@@ -0,0 +1,1352 @@
+}
+
+
+std::vector<TxEntry::TxEntryRef> TxGraph::GetDescendants(const std::vector<TxEntry::TxEntryRef>& txs) const
+{
+    std::vector<TxEntry::TxEntryRef> result{}, work_queue{};
+
+    LOCK(cs);
+
+    WITH_FRESH_EPOCH(m_epoch);
+    for (auto tx : txs) {
+        if (!visited(tx)) {
+            work_queue.push_back(tx);
+        }
+    }
+
+    while (!work_queue.empty()) {
+        auto it = work_queue.back();
+        work_queue.pop_back();
+        result.push_back(it);
+        for (auto& child: it.get().GetTxEntryChildren()) {
+            if (!visited(child.get())) {
+                work_queue.push_back(child);
+            }
+        }
+    }
+    return result;
+}
+
+std::vector<TxEntry::TxEntryRef> TxGraph::GetAncestors(const std::vector<TxEntry::TxEntryRef>& parents) const
+{
+    std::vector<TxEntry::TxEntryRef> result{}, work_queue{};
+
+    LOCK(cs);
+
+    WITH_FRESH_EPOCH(m_epoch);
+    for (auto p : parents) {
+        visited(p);
+        work_queue.push_back(p);
+    }
+
+    while (!work_queue.empty()) {
+        auto it = work_queue.back();
+        work_queue.pop_back();
+        result.push_back(it);
+        for (auto& parent: it.get().parents) {
+            if (!visited(parent.get())) {
+                work_queue.push_back(parent);
+            }
+        }
+    }
+    return result;
+}
+
+std::vector<TxEntry::TxEntryRef> TxGraph::GatherAllClusterTransactions(const std::vector<TxEntry::TxEntryRef> &txs) const
+{
+    LOCK(cs);
+
+    WITH_FRESH_EPOCH(m_epoch);
+    std::vector<TxEntry::TxEntryRef> result;
+    for (auto tx : txs) {
+        if (!visited(tx.get().m_cluster)) {
+            for (auto &chunk : tx.get().m_cluster->m_chunks) {
+                for (auto &txentry : chunk.txs) {
+                    result.push_back(txentry);
+                }
+            }
+        }
+    }
+    return result;
+}
+
+TxGraphCluster* TxGraph::AssignTxGraphCluster()
+{
+    auto new_cluster = std::make_unique<TxGraphCluster>(m_next_cluster_id++, this);
+    TxGraphCluster* ret = new_cluster.get(); // XXX No one is going to like this.
+    m_cluster_map[new_cluster->m_id] = std::move(new_cluster);
+    return ret;
+}
+
+// When transactions are removed from a cluster, the cluster might get split
+// into smaller clusters.
+void TxGraph::RecalculateTxGraphClusterAndMaybeSort(TxGraphCluster *cluster, bool sort)
+{
+    // TODO: if the common case involves no cluster splitting, can we short
+    // circuit the work here somehow?
+
+    // Wipe cluster assignments.
+    std::vector<TxEntry::TxEntryRef> txs;
+    for (auto& chunk : cluster->m_chunks) {
+        for (auto& txentry: chunk.txs) {
+            txentry.get().m_cluster = nullptr;
+            txs.push_back(txentry);
+        }
+    }
+    cluster->Clear();
+
+    // The first transaction gets to stay in the existing cluster.
+    bool first = true;
+    for (auto& txentry : txs) {
+        if (txentry.get().m_cluster == nullptr) {
```

Comment:
Isn't this trivially true from this: https://github.com/bitcoin/bitcoin/pull/28676/files#diff-7dcb5aa190f52e5da3aca406127995f65dab3ee9d33ab68130cb3460d2298c7eR498

Reply:
See line 528 below: https://github.com/sdaftuar/bitcoin/blob/c49e0444e5c474d5e1ac0af89e9bc27958f3ed31/src/kernel/txgraph.cpp#L528